### ISSUE TITLE
Remove unwanted KubeArmor service running in kube-system namespace

### ISSUE THREAD

**Issue by viveksahu26** [kubearmor/KubeArmor#667]
## Feature Request

**Short Description**
Remove KubeArmor services running in kube-system namespace.

**Is your feature request related to a problem? Please describe the use case.**
Basically, these are auto-generated services by kubebuilder  and currently they are of no use.
To see those service,
```
[linuzz@fedora ~]$ sudo microk8s kubectl get svc -n kube-system | grep kubearmor
kubearmor                                       ClusterIP   10.152.183.209   <none>        32767/TCP                8d
kubearmor-policy-manager-metrics-service        ClusterIP   10.152.183.222   <none>        8443/TCP                 8d
kubearmor-host-policy-manager-metrics-service   ClusterIP   10.152.183.249   <none>        8443/TCP                 8d

```
**Describe the solution you'd like**
Clean up those services.

**Describe alternatives you've considered**

**Comment by viveksahu26** [comment]
Hey @daemon1024 , since I don't have permission to add `cleanup` tag over `enhancement` tag. So,  can you do it.

**Comment by daemon1024** [comment]
> I don't have permission to add cleanup tag over enhancement tag. So, can you do it.

`enhancement` should be fine.

>kubearmor                                       ClusterIP   10.152.183.209   <none>        32767/TCP          

This is an important service, we connect to this for logs/telemetry.

>kubearmor-policy-manager-metrics-service        ClusterIP   10.152.183.222   <none>        8443/TCP                 8d
kubearmor-host-policy-manager-metrics-service   ClusterIP   10.152.183.249   <none>        8443/TCP     

I would like to understand what potential usecase would this have before proceeding to remove it. We might consider leveraging those usecases as well if they deem fit.

**Comment by viveksahu26** [comment]
Hey @daemon1024 , as you said that below service 

> kubearmor ClusterIP 10.152.183.209 32767/TCP

is used in providing logs and telemetry.

But, I deleted all 3 services from my cluster.

> kubearmor ClusterIP 10.152.183.209 32767/TCP
> kubearmor-policy-manager-metrics-service ClusterIP 10.152.183.222 8443/TCP 8d
> kubearmor-host-policy-manager-metrics-service ClusterIP 10.152.183.249 8443/TCP

Then also `karmor cli` provides me logs and telemetry.  I don't understand why so ?

At present time these all are services running in my cluster
```
kubectl get svc   -n kube-system
NAME                        TYPE        CLUSTER-IP       EXTERNAL-IP   PORT(S)                  AGE
kube-dns                    ClusterIP   10.152.183.10    <none>        53/UDP,53/TCP,9153/TCP   17d
metrics-server              ClusterIP   10.152.183.170   <none>        443/TCP                  17d
kubernetes-dashboard        ClusterIP   10.152.183.171   <none>        443/TCP                  17d
dashboard-metrics-scraper   ClusterIP   10.152.183.70    <none>        8000/TCP                 17d

```

**Comment by daemon1024** [comment]
The `kubearmor svc` here is part of the https://github.com/kubearmor/kubearmor-relay-server, if you check the selector it would have `kubearmor-app: kubearmor-relay`

> But, I deleted all 3 services from my cluster.
Then also karmor cli provides me logs and telemetry. I don't understand why so ?

What you are doing right now is essentially connecting to `kubearmor` directly. This only works because you are using kArmor on the same node. Also if there are multiple kubearmor in your cluster ( due to multiple nodes) you won't be getting the aggregated feed. To know more about the relay server you can check out the [README](https://github.com/kubearmor/kubearmor-relay-server).

**Comment by viveksahu26** [comment]
Yeah, that could be the reason. Ok, I will also try to check it by installing both(k8s cluster and karmor cli) of them in diffferent nodes.

**Comment by viveksahu26** [comment]
Basically there are 3 deployments, 
```
kubectl get deploy -n kube-system | grep kubearmor
kubearmor-relay                 1/1     1            1           17d
kubearmor-policy-manager        1/1     1            1           17d
kubearmor-host-policy-manager   1/1     1            1           17d

```
> kubearmor-relay and 
> kubearmor ClusterIP 10.152.183.209 32767/TCP

are related to each other.

Similarly, 
> kubearmor-policy-manager deployment and 
> kubearmor-policy-manager-metrics-service ClusterIP 10.152.183.222 8443/TCP 8d     service 

are related to each other. 

And, 
> kubearmor-host-policy-manager         deployment
> kubearmor-host-policy-manager-metrics-service ClusterIP 10.152.183.249 8443/TCP            service 

aree related to each other.

**Comment by viveksahu26** [comment]
What does deployment `kubearmor-host-policy-manager ` and `kubearmor-policy-manager ` does. What is their use?

**Comment by viveksahu26** [comment]
And both deployment uses kubebuilder images, i.e. `gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0`
https://github.com/kubearmor/KubeArmor/blob/4faeb7f825c175ac96600c1ba0b58085b2765bd0/deployments/microk8s/kubearmor.yaml#L215

**Comment by daemon1024** [comment]
> What does deployment `kubearmor-host-policy-manager ` and `kubearmor-policy-manager ` does. What is their use?

They contain controllers for policy validation. Ref https://github.com/kubearmor/KubeArmor/blob/main/pkg/KubeArmorPolicy/controllers/kubearmorpolicy_controller.go

> And both deployment uses kubebuilder images, i.e. `gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0`

One of the image is rbac-proxy, their main image contains the CRD validation controller.
https://github.com/kubearmor/KubeArmor/blob/4faeb7f825c175ac96600c1ba0b58085b2765bd0/deployments/microk8s/kubearmor.yaml#L232

**Comment by viveksahu26** [comment]
Hey @daemon1024 , I think controllers defined for deployments `kubearmor-host-policy-manager` and `kubearmor-policy-manager` will handle policy validation as well as internal logic. There is no need of services corresponding to  those deployment.
I have also tried to proof it.
Basically I ran the policy in presence of deployment  `kubearmor-host-policy-manager` and its service `kubearmor-policy-manager-metrics-service`.
policy.yaml
```
apiVersion: security.kubearmor.com/v1
kind: KubeArmorPolicy
metadata:
  name: ksp-group-1-proc-path-block
  namespace: multiubuntu
spec:
  severity: 5
  message: "block the sleep command"
  selector:
    matchLabels:
      group: group-1
  process:
    matchPathss:
      - path: /bin/sleep # try sleep 1 (permission denied)
  action:
    Block
```
Then it return below errors(because the above policy has `matchPathss` instead of `matchPath`:
```
linuzz@hp:~/go/src/github.com/kubearmor/KubeArmor/examples$ kubectl apply -f multiubuntu/security-policies/ksp-group-1-proc-path-block.yaml 
error: error validating "multiubuntu/security-policies/ksp-group-1-proc-path-block.yaml": error validating data: ValidationError(KubeArmorPolicy.spec.process): unknown field "matchPathss" in com.kubearmor.security.v1.KubeArmorPolicy.spec.process; if you choose to ignore these errors, turn validation off with --validate=false
```

After that I deleted the service `kubearmor-policy-manager-metrics-service`, and repeated the above scenario and got same errors as above.

So, I  don't think there is any role of the kubearmor-policy-manager-metrics-service` service. 

What do you think?

**Comment by viveksahu26** [comment]
@daemon1024 , any thoughts  ??

**Comment by daemon1024** [comment]
@viveksahu26 The Validation that you tried out right now is the validation performed by k8s based on the CRD specification and not by the controller.

You can see here what the controller validates for 
https://github.com/kubearmor/KubeArmor/blob/bf6087fd0bc04315ccc3afd9babd72f215ec77ac/pkg/KubeArmorPolicy/controllers/kubearmorpolicy_controller.go#L123
Like here it checks if we mandate an allow action to be used with owner only


Thanks a lot for spending time to analyse things here. Really appreciate it.

**Comment by viveksahu26** [comment]
> @viveksahu26 The Validation that you tried out right now is the validation performed by k8s based on the CRD specification and not by the controller.

Yes, it's CRD specs.

**Comment by viveksahu26** [comment]
@daemon1024 , Do you want me to research more on above services functionality. Or should I remove it. As I couldn't see any role of those services.

**Comment by viveksahu26** [comment]
Found an interesting [docs](https://github.com/brancz/kube-rbac-proxy)  on image(`gcr.io/kubebuilder/kube-rbac-proxy:v0.5.0`) used by deployments(`kubearmor-host-policy-manager` and `kubearmor-policy-manager`). kube-rbac-proxy is a small HTTP proxy for a single upstream, that can perform RBAC authorization against the Kubernetes API using [SubjectAccessReview](https://kubernetes.io/docs/reference/access-authn-authz/authorization/).[SubjectAccessReview](https://kubernetes.io/docs/reference/access-authn-authz/authorization/).

**Comment by viveksahu26** [comment]
@daemon1024 , Any thoughts on this ??

**Comment by Aryan-sharma11** [comment]
Closing this as a stale issue.
